packages/core/test/userState.test.ts

```diff
@@ -0,0 +1,293 @@
+
+        const epoch = await userState.getUnirepStateCurrentEpoch()
+        const newPosRep = 10
+        const newNegRep = 5
+        const newGraffiti = 1294194
+        // now submit the attestation from the attester
+        await unirepContract
+            .connect(attester)
+            .submitAttestation(
+                epoch,
+                epochKey,
+                newPosRep,
+                newNegRep,
+                newGraffiti
+            )
+            .then((t) => t.wait())
+        await userState.waitForSync()
+        // now commit the attetstations
+        await unirepContract
+            .connect(accounts[5])
+            .buildHashchain(attester.address, epoch)
+            .then((t) => t.wait())
+        const hashchain = await unirepContract.attesterHashchain(
+            attester.address,
+            epoch,
+            0
+        )
+        const { publicSignals, proof } =
+            await userState.genAggregateEpochKeysProof(
+                hashchain.epochKeys,
+                hashchain.epochKeyBalances,
+                hashchain.index,
+                epoch
+            )
+        await unirepContract
+            .connect(accounts[5])
+            .processHashchain(publicSignals, proof)
+            .then((t) => t.wait())
+        await userState.stop()
+    })
+
+    it('ust proof', async () => {
+        const accounts = await ethers.getSigners()
+        const attester = accounts[1]
+        const attesterId = BigInt(attester.address)
+        const id = new ZkIdentity()
+        const userState = await genUserState(
+            ethers.provider,
+            unirepContract.address,
+            id,
+            attesterId
+        )
+
+        {
+            const { publicSignals, proof } =
+                await userState.genUserSignUpProof()
+            await unirepContract
+                .connect(attester)
+                .userSignUp(publicSignals, proof)
+                .then((t) => t.wait())
+        }
+        const oldEpoch = await userState.latestTransitionedEpoch()
+        // epoch transition
+        await ethers.provider.send('evm_increaseTime', [EPOCH_LENGTH])
+        await ethers.provider.send('evm_mine', [])
+        {
+            await userState.waitForSync()
+            const { publicSignals, proof } =
+                await userState.genUserStateTransitionProof()
+            // submit it
+            await unirepContract
+                .connect(accounts[4])
+                .userStateTransition(publicSignals, proof)
+                .then((t) => t.wait())
+        }
+        await userState.waitForSync()
+        const newEpoch = await userState.latestTransitionedEpoch()
+        expect(newEpoch).equal(oldEpoch + 1)
+        await userState.stop()
+    })
+
+    // TODO: should fix reputation proof first
+    // it('reputation proof', async () => {
+    //     const accounts = await ethers.getSigners()
+    //     const attester = accounts[1]
```

Comment:
We have a functioning reputation proof. I think we should separate the spending proof and the rep proof though.